caom2-pkg-server/src/main/java/ca/nrc/cadc/caom2/pkg/Caom2PackageRunner.java

```diff
@@ -0,0 +1,272 @@
+*
+*  $Revision: 5 $
+*
+************************************************************************
+*/
+
+package ca.nrc.cadc.caom2.pkg;
+
+import org.opencadc.pkg.server.PackageItem;
+import org.opencadc.pkg.server.PackageRunner;
+import org.opencadc.pkg.server.PackageRunnerException;
+
+import ca.nrc.cadc.auth.AuthMethod;
+import ca.nrc.cadc.auth.AuthenticationUtil;
+import ca.nrc.cadc.caom2.Artifact;
+import ca.nrc.cadc.caom2.ObservationURI;
+import ca.nrc.cadc.caom2.PlaneURI;
+import ca.nrc.cadc.caom2.ProductType;
+import ca.nrc.cadc.caom2.PublisherID;
+import ca.nrc.cadc.caom2.artifact.resolvers.CaomArtifactResolver;
+import ca.nrc.cadc.caom2ops.ArtifactQueryResult;
+import ca.nrc.cadc.caom2ops.CaomTapQuery;
+import ca.nrc.cadc.caom2ops.ServiceConfig;
+import ca.nrc.cadc.log.WebServiceLogInfo;
+import ca.nrc.cadc.net.ResourceNotFoundException;
+import ca.nrc.cadc.util.StringUtil;
+import ca.nrc.cadc.uws.ErrorSummary;
+import ca.nrc.cadc.uws.ErrorType;
+import ca.nrc.cadc.uws.ParameterUtil;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URL;
+import java.net.URISyntaxException;
+import java.security.AccessControlContext;
+import java.security.AccessController;
+import java.security.cert.CertificateException;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+import javax.security.auth.Subject;
+import org.apache.log4j.Logger;
+
+
+public class Caom2PackageRunner extends PackageRunner {
+    private static final Logger log = Logger.getLogger(Caom2PackageRunner.class);
+
+    private final URI tapID;
+    private List<String> idList;
+    private String packageName;
+
+    public Caom2PackageRunner()
+    { 
+        ServiceConfig sc = new ServiceConfig();
+        this.tapID = sc.getTapServiceID();
+    }
+
+    @Override
+    protected String getPackageName() {
+        return this.packageName;
+    }
+
+
+    @Override
+     public Iterator<PackageItem> getItems() throws IOException, PackageRunnerException {
+
+        // obtain credentials from CDP if the user is authorized
+        AccessControlContext accessControlContext = AccessController.getContext();
+        Subject subject = Subject.getSubject(accessControlContext);
+        AuthMethod authMethod = AuthenticationUtil.getAuthMethod(subject);
+
+        String runID = this.job.getID();
+        if (this.job.getRunID() != null) {
+            runID = this.job.getRunID();
+        }
+
+        List<String> idList = ParameterUtil.findParameterValues("ID", this.job.getParameterList());
+
+        CaomTapQuery query = new CaomTapQuery(tapID, runID);
+
+        CaomArtifactResolver artifactResolver = new CaomArtifactResolver();
+        artifactResolver.setAuthMethod(authMethod); // override auth method for proxied calls
+        artifactResolver.setRunID(runID);
+
+        List<PackageItem> packageItems = new ArrayList<PackageItem>();
```

Comment:
Finally,  the previous impl was doing output inside the loop over input ID values (which are already loaded into memory) so it didn't have an additional memory footprint. Constructing the entire List<PackageItem> will add a lot of latency and memory consumption... so it's not scalable.

Two options:
1. non-scalable: impose a limit on the number of input ID and throw if there are too many; since we don't know how many artifacts each plane has, this isn't ever a great solution
2. return a dynamic iterator that uses CaomTapQuery internally to the the next batch of artifacts; this would use a "look-ahead" style with two internal iterations: outer Iterator<URI> and inner Iterator<Artifact>.

Once you get it working, option 2 is much better and very scalable. Most of the checked exceptions related to invalid input are outside if we go with `initPackage()` described in another comment. This scalable dynamic iteterator style will definitely be needed in a VOSspacePackageRunner so working it out here could be beneficial.

Reply:
For the first pass, leaving this as is, as implementing 2. is preferred and can be done in a separate step.